src/lib/core/ReferenceCounted.h

```diff
@@ -51,6 +51,10 @@ class ReferenceCounted
     /** Adds one to the usage count of this class */
     Subclass * Retain()
     {
+        if (kInitRefCount && mRefCount == 0)
```

Comment:
Consider using a compile time static_assert to check that `kInitRefCount` is not 0, if that is appropriate for this template.

Reply:
Unfortunately, this code purposefully instantiates the template with  kInitRefCount=0: https://github.com/project-chip/connectedhomeip/blob/4b9ead687c2b3eed1e430c1314b3a16bde211fc8/src/transport/UnauthenticatedSessionTable.h#L48

This new check has to ignore usages where that's the case.